File: branches/Medusa/Source/Fantasy.Jobs/TrackProviderService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Fantasy.Tracking;
using Fantasy.ServiceModel;


namespace Fantasy.Jobs
{
    public class TrackProviderService : AbstractService, ITrackProvider, IJobEngineEventHandler, IDisposable 

    {

        ITrackProvider _provider;

        private Dictionary<string, object> _initValues = new Dictionary<string, object>();



        public override void InitializeService()
        {
            IJobEngine engine = this.Site.GetService<IJobEngine>();
            engine.AddHandler(this);
            base.InitializeService();
        }

        public Guid Id
        {
            get { return _provider != null ? _provider.Id : this.Site.GetService<IJobEngine>().JobId; }
        }

        public string Name
        {
            get { return _provider.Name; }
        }

        public string Category
        {
            get { return _provider != null ? _provider.Category : string.Empty; }
        }

        public object this[string name]
        {
            get
            {
                return _provider != null ? _provider[name] : _initValues.GetValueOrDefault(name, null); 
            }
            set
            {
                if (_provider != null)
                {
                    _provider[name] = value;
                }
                else
                {
                    _initValues[name] = value;
                }
            }
        }

        public string[] PropertyNames
        {
            get { return _provider != null ? _provider.PropertyNames : _initValues.Keys.ToArray(); }
        }

        public TrackFactory Connection
        {
            get { return _provider != null ? _provider.Connection : null; }
        }

        

        #region IJobEngineEventHandler Members

        void IJobEngineEventHandler.HandleStart(IJobEngine sender)
        {
           
        }

        void IJobEngineEventHandler.HandleResume(IJobEngine sender)
        {
           
        }

        void IJobEngineEventHandler.HandleExit(IJobEngine sender, JobExitEventArgs e)
        {
            

           
        }
 
        void IJobEngineEventHandler.HandleLoad(IJobEngine sender)
        {
            Job job = this.Site.GetService<Job>();

            string name = job.GetProperty("name");
            if (string.IsNullOrWhiteSpace(name))
            {
                name = job.TemplateName + DateTime.Now.ToString(); 
            }

            TrackFactory cnnt = new TrackFactory();
            _provider = cnnt.CreateProvider(job.ID, name, "Jobs." + job.TemplateName, _initValues);
        }

        #endregion

        #region IDisposable Members

        public void Dispose()
        {
        }

        #endregion
    }
}

File: Fantasy.Jobs/src/fantasy/jobs/TrackProviderService.java
package fantasy.jobs;

import java.rmi.RemoteException;
import java.util.*;

import fantasy.*;
import fantasy.tracking.*;
import fantasy.servicemodel.*;

public class TrackProviderService extends AbstractService implements ITrackProvider, IJobEngineEventHandler, IDisposable

{

	/**
	 * 
	 */
	private static final long serialVersionUID = -6317840824010522511L;



	public TrackProviderService() throws RemoteException {
		super();
		
	}


	private ITrackProvider _provider;

	private java.util.HashMap<String, Object> _initValues = new java.util.HashMap<String, Object>();



	@Override
	public void initializeService() throws Exception
	{
		IJobEngine engine = this.getSite().getRequiredService(IJobEngine.class);
		engine.AddHandler(this);
		super.initializeService();
	}
	
	@Override 
	public void uninitializeService() throws Exception
	{
		super.uninitializeService();
		this.dispose();
	}

	public final UUID getId()
	{
		try
		{
		return _provider != null ? _provider.getId() : this.getSite().getRequiredService(IJobEngine.class).getJobId();
		}
		catch(Exception e)
		{
			return null;
		}
	}

	public final String getName()
	{
		return _provider.getName();
	}

	public final String getCategory()
	{
		return _provider != null ? _provider.getCategory() : "";
	}

	public final Object getProperty(String name)
	{
		return _provider != null ? _provider.getProperty(name) :  _initValues.get(name);
	}
	
	@SuppressWarnings("unchecked")
	@Override
	public <T> T getProperty(String name, T defaultValue) {
		Object rs = this.getProperty(name);
		
		return rs != null ? (T)rs : defaultValue;
	}
	
	public final void setProperty(String name, Object value)
	{
		if (_provider != null)
		{
			_provider.setProperty(name, value);
		}
		else
		{
			_initValues.put(name, value);
		}
	}

	public final String[] getPropertyNames()
	{
		return _provider != null ? _provider.getPropertyNames() : _initValues.keySet().toArray(new String[0]);
	}

	


	public void HandleStart(IJobEngine sender)
	{

	}

	public void HandleResume(IJobEngine sender)
	{

	}

	public void HandleExit(IJobEngine sender, JobExitEventArgs e)
	{



	}

	
	private ITrackManager _manager;
	
	public void HandleLoad(IJobEngine sender) throws Exception
	{
		IJob job = this.getSite().getService(IJob.class);

		String name = job.GetProperty("name");
		if (StringUtils2.isNullOrWhiteSpace(name))
		{
			name = job.getTemplateName() + new java.util.Date().toString();
		}

		_manager = TrackManagerFactory.createTrackManager();

		_provider = _manager.getProvider(job.getID(), name, "Jobs." + job.getTemplateName(), _initValues);
	}


	public final void dispose()
	{
		if(_provider != null)
		{
		    _provider.dispose();
		}
	}

}